File: Arrays and Hashing/top-k-frequent-elements.py
# time O(n) space O(n)
class Solution:
    def topKFrequent(self, nums, k):
        hash_map = {}
        bucket = [[] for i in range(len(nums) + 1)]
        result = []
        for num in nums:
            hash_map[num] = hash_map.get(num, 0) + 1
        for key, val in hash_map.items():
            bucket[val].append(key)
        for i in range(len(bucket) - 1, 0, -1):
            for n in bucket[i]:
                result.append(n)
                if len(result) == k:
                    return result

File: Arrays and Hashing/top-k-frequent-elements.js
// time O(n) space O(n)
var topKFrequent = function (nums, k) {
  let hashMap = new Map();
  let bucket = Array(nums.length);
  let result = [];
  for (let num of nums) {
    let val = hashMap.get(num) + 1 || 1;
    hashMap.set(num, val);
  }
  for (let [key, val] of hashMap.entries()) {
    bucket[val] = (bucket[val] || new Set()).add(key);
  }

  for (let i = bucket.length - 1; i >= 0; i--) {
    if (bucket[i] !== undefined) {
      result.push(...bucket[i]);
    }
    if (result.length >= k) {
      return result;
    }
  }
};
